### ISSUE TITLE
Apostrophes in contractions are not converted to right single quote

### ISSUE THREAD

**Issue by bryanforbes** [yuin/goldmark#39]
For the following text:

```
I'm going to see my mother. She's very nice.
```

Currently, `'` is not converted to `&rsquo;` for contractions when the typography extension is enabled, but smartypants does. I would expect the output to be:

```
I&rsquo;m going to see my mother. She&rsquo;s very nice.
```

**Comment by sungchang-ha** [comment]
It appears that the single quote (identical to apostrophe in Markdown input) is not converted as intended in the following case.

Source text (note the period at the end):

`'sample'.`

Expected:

`&lsquo;sample&rsquo;.`

Actual result:

`&lsquo;sample'.`

Let me know if I should create a separate issue for this.

**Comment by yuin** [comment]
@bryanforbes Typographer extension is just a typographer extension is not SmartyPants exntension. Typographer extension does not support your case.

@sungchang-ha I can not reproduce this. `'sample'.` is converted to `&lsquo;sample&rsquo;.` in my environment.

**Comment by sungchang-ha** [comment]
@yuin I should've provided the test string as-is, which I edited to no avail. 

It appears that the combination of inline styles (bold, italic, etc.) and the adjacent punctuation mark causes the issue. Test environment is Goldmark 1.1.8 and the made-up test string is shown below:

* Markdown: `'*en-US*'/'**javascript**', '*sample*'.`
* Expected: `&lsquo;<em>en-US</em>&rsquo;/&lsquo;<strong>javascript</strong>&rsquo;, &lsquo;<em>sample<em>&rsquo;.`
* Current: `&lsquo;<em>en-US</em>'/'<strong>javascript</strong>', &lsquo;<em>sample</em>'.`

Removing either of the inline style or the adjacent punctuation mark makes it rendered as quotes.

**Comment by yuin** [comment]
@sungchang-ha Reasonable. Typographer extension is implemented with same algorithm as [Emphasis and strong emphasis defined in CommonMark](https://spec.commonmark.org/0.29/#emphasis-and-strong-emphasis). Left flanking is converted into `&lsquo;` and Right flanking is converted into `&rsquo;` . This is a limitation in this algorithm.

I'm not usually writing English, so this functionalities has low priority for me. 
PR with enhancement of typographer is welcome.

**Comment by bryanforbes** [comment]
> Typographer extension does not support your case.

Do you mean that converting apostrophes to single right quote will not be supported ever, or that converting them is currently unsupported?

**Comment by yuin** [comment]
@bryanforbes Converting them is currently unsupported. As wrote above, I'm not usually writing English and honestly say I never used this functionality.

I do not have so much time for this project, so I'm going to use the time for better supporting CJK environment instead of typographer that I have never used.

I welcome PR that enhance typographer(or add complete SmartyPants extension).

**Comment by rgov** [comment]
Before trying to fix the cases where Typographer doesn't work properly, is an inline parser the correct model for this transformation? To fix this, we would need to have the plain text without formatting.

**Comment by yuin** [comment]
@rgov 
Probably not. There's no efficient and 'works perfectly' model in goldmark right now for this.

goldmark is an extensible markdown parser, so we can not know about information of user-assigned extension(ast nodes, syntax, html output, etc).

Furthermore, users can remove or replace built-in parsers and renderers. If users remove a built-in emphasis parser, your example should yield `**this is** a test` .  If users replace built-in emphasis renderer with their own renderer that render `**` as `IMPORTANT{}`, your example should yield `IMPORTANT{this is} a test`.
These are extreme cases though.  

I think parsing html output is the only way that you **perfectly** apply smartypants substitutions. 
If we give **special** treatment for smartypants, we are probably able to find more efficient way. But I'm not willing to give such treatment for English specific functionalities like smartypants.

**Comment by yuin** [comment]
I've tried to fix this in 3393022.
But I'm not familiar with English, Could anyone confirm whether this is fixed.
